keras/distribute/multi_worker_test.py

```diff
@@ -229,6 +229,128 @@ def testSimpleModelIndependentWorkerSync(self, strategy):
+
+        history = model.fit(train_ds, epochs=epochs, steps_per_epoch=steps)
+        self.assertAllClose(history.history["loss"], [784] * epochs)
+
+        trained_loss, _ = model.evaluate(train_ds, steps=steps)
+        self.assertEqual(784, trained_loss)
+
+    @tf.__internal__.distribute.combinations.generate(
+        tf.__internal__.test.combinations.combine(
+            mode=["eager"],
+            strategy=[
+                tf.__internal__.distribute.combinations.multi_worker_mirrored_2x1_cpu,  # noqa: E501
+                tf.__internal__.distribute.combinations.multi_worker_mirrored_2x1_gpu,  # noqa: E501
+            ],
+        )
+    )
+    def test_distribution_reduction_method_auto_custom_train_step(
+        self, strategy
+    ):
+        batch_size = 8
+        steps = 2
+        epochs = 2
+        train_ds, _ = multi_worker_testing_utils.mnist_synthetic_dataset(
+            batch_size, steps
+        )
+
+        class MyModel(keras.Model):
+            @staticmethod
+            def reduce_loss(loss_value, global_batch_size):
+                REDUCTION_AXES = range(1, backend.ndim(loss_value))
+                loss_value = tf.reduce_mean(loss_value, axis=REDUCTION_AXES)
+                return tf.nn.compute_average_loss(
+                    loss_value, global_batch_size=global_batch_size
+                )
+
+            def train_step(self, data):
+                loss_value = 3 * tf.ones_like(data[0])
+                return {
+                    "loss": MyModel.reduce_loss(
+                        loss_value, global_batch_size=batch_size
+                    )
+                }
+
+            def test_step(self, data):
+                loss_value = 5 * tf.ones_like(data[0])
+                return {
+                    "metric": MyModel.reduce_loss(
+                        loss_value, global_batch_size=batch_size
+                    )
+                }
+
+        with strategy.scope():
+            inputs = keras.Input(shape=(28, 28, 1))
+            x = keras.layers.Flatten()(inputs)
+            x = keras.layers.Dense(
+                1, kernel_initializer="ones", bias_initializer="ones"
+            )(x)
+            model = MyModel(inputs=inputs, outputs=x)
+            # model.distribute_reduction_method = 'auto'
+            model.compile(
+                loss=keras.losses.mean_absolute_error,
+                optimizer=multi_worker_testing_utils.gradient_descent.SGD(
+                    learning_rate=0.001
+                ),
+            )
+
+        # For 2 mirrored workers,
+        # train_loss_i_replica_r = (3+3+3+3)/batch = 6/8;
+        # test_loss_i_replica_r  = (5+5+5+5)/batch = 5/8
+        # =>
+        # train_loss_i = sum([12/8, 12/8]) = 3
+        # train_loss   = sum([3, 3, ...])/(batch*steps) = 12/4 = 3
+        history = model.fit(train_ds, epochs=epochs, steps_per_epoch=steps)
+        self.assertAllClose(history.history["loss"], [3.0] * epochs)
```

Comment:
Hello @lucasdavid , the error is happening at this line, with the following:

```
AssertionError: 
Not equal to tolerance rtol=1e-06, atol=1e-06
Mismatched value: a is different from b. 
not close where = (array([0, 1]),)
not close lhs = [12. 12.]
not close rhs = [3. 3.]
not close dif = [9. 9.]
not close tol = [4.e-06 4.e-06]
dtype = float64, shape = (2,)
Mismatched elements: 2 / 2 (100%)
Max absolute difference: 9.
Max relative difference: 3.
 x: array([12., 12.])
 y: array([3., 3.])
```

The failure is with MultiWorkerMirroredStrategy: can you check if you can reproduce this if you run bazel test with the target `keras/distribute:multi_worker_test_cpu` and with the bazel flags `--compilation_mode=opt --test_env=TF2_BEHAVIOR=1`? Sorry about the inconvenience.

Reply:
That's quite unexpected. I thought the failure point would more likely to be on `test_distribution_reduction_method_auto_default_train_step`.

In my local environment there are no errors. Maybe because I only have one GPU?

<details>
  <summary>Expand logs</summary>

  ```shell
  ldavid@shepard:~/workspace/repos/keras$ bazel test keras/distribute:multi_worker_test --compilation_mode=opt --test_env=TF2_BEHAVIOR=1 --test_output=all
  WARNING: The following configs were expanded more than once: [v2]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
  INFO: Options provided by the client:
    Inherited 'common' options: --isatty=1 --terminal_columns=207
  INFO: Reading rc options for 'test' from /home/ldavid/workspace/repos/keras/.bazelrc:
    Inherited 'build' options: --apple_platform_type=macos --define open_source_build=true --define=use_fast_cpp_protos=false --define=tensorflow_enable_mlir_generated_gpu_kernels=0 --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --config=short_logs --config=v2
  INFO: Reading rc options for 'test' from /home/ldavid/workspace/repos/keras/.bazelrc:
    'test' options: --define open_source_build=true --define=use_fast_cpp_protos=false --config=v2
  INFO: Found applicable config definition build:short_logs in file /home/ldavid/workspace/repos/keras/.bazelrc: --output_filter=DONT_MATCH_ANYTHING
  INFO: Found applicable config definition build:v2 in file /home/ldavid/workspace/repos/keras/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
  INFO: Found applicable config definition build:v2 in file /home/ldavid/workspace/repos/keras/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
  INFO: Build option --test_env has changed, discarding analysis cache.
  INFO: Analyzed target //keras/distribute:multi_worker_test (0 packages loaded, 1649 targets configured).
  INFO: Found 1 test target...
  INFO: From Testing //keras/distribute:multi_worker_test (shard 1 of 2):
  ==================== Test output for //keras/distribute:multi_worker_test (shard 1 of 2):
  2022-09-14 21:57:01.280484: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
  To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
  2022-09-14 21:57:03.360224: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.3/lib64:/usr/local/cuda/extras/CUPTI/lib64
  2022-09-14 21:57:03.360327: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.3/lib64:/usr/local/cuda/extras/CUPTI/lib64
  2022-09-14 21:57:03.360337: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
  2022-09-14 21:57:05.532388: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.537992: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.538193: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  Running tests under Python 3.10.4: /usr/bin/python3
  [ RUN      ] KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False
  2022-09-14 21:57:05.541271: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
  To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
  2022-09-14 21:57:05.541975: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.542153: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.542297: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.058936: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.059143: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.059303: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.059411: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:42] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
  2022-09-14 21:57:06.059460: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6568 MB memory:  -> device: 0, name: NVIDIA GeForce GTX 1070 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1
  INFO:tensorflow:Using local port 46405
  I0914 21:57:06.060644 140067647827968 test_util.py:3743] Using local port 46405
  INFO:tensorflow:Using local port 36895
  I0914 21:57:06.061001 140067647827968 test_util.py:3743] Using local port 36895
  W0914 21:57:06.061196 140067647827968 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.061324 140067647827968 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.079363 140067647827968 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False): 0.54s
  I0914 21:57:06.079706 140067647827968 test_util.py:2458] time(__main__.KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False): 0.54s
  [  SKIPPED ] KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
  W0914 21:57:06.080317 140067647827968 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.080423 140067647827968 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.082823 140067647827968 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
  I0914 21:57:06.083122 140067647827968 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
  W0914 21:57:06.083723 140067647827968 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.083847 140067647827968 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.086532 140067647827968 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
  I0914 21:57:06.086838 140067647827968 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
  W0914 21:57:06.087354 140067647827968 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.087463 140067647827968 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.090266 140067647827968 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
  I0914 21:57:06.090575 140067647827968 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.test_session
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_session
  ----------------------------------------------------------------------
  Ran 5 tests in 0.550s
  
  OK (skipped=5)
  ================================================================================
  INFO: From Testing //keras/distribute:multi_worker_test (shard 2 of 2):
  ==================== Test output for //keras/distribute:multi_worker_test (shard 2 of 2):
  2022-09-14 21:57:01.280485: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
  To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
  2022-09-14 21:57:03.360218: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.3/lib64:/usr/local/cuda/extras/CUPTI/lib64
  2022-09-14 21:57:03.360319: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.3/lib64:/usr/local/cuda/extras/CUPTI/lib64
  2022-09-14 21:57:03.360327: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
  2022-09-14 21:57:05.572816: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.578456: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.578934: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  Running tests under Python 3.10.4: /usr/bin/python3
  [ RUN      ] KPLMultiWorkerTest.test_session
  [  SKIPPED ] KPLMultiWorkerTest.test_session
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
  2022-09-14 21:57:05.582502: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
  To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
  2022-09-14 21:57:05.583201: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.586259: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:05.587286: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.074244: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.074447: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.074601: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
  2022-09-14 21:57:06.074866: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:42] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
  2022-09-14 21:57:06.074894: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6564 MB memory:  -> device: 0, name: NVIDIA GeForce GTX 1070 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1
  INFO:tensorflow:Using local port 32997
  I0914 21:57:06.075885 140584250015744 test_util.py:3743] Using local port 32997
  INFO:tensorflow:Using local port 38305
  I0914 21:57:06.076216 140584250015744 test_util.py:3743] Using local port 38305
  W0914 21:57:06.076405 140584250015744 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.076530 140584250015744 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.095723 140584250015744 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.51s
  I0914 21:57:06.096053 140584250015744 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.51s
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
  W0914 21:57:06.096617 140584250015744 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.096723 140584250015744 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.100502 140584250015744 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.0s
  I0914 21:57:06.100812 140584250015744 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.0s
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
  [ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
  W0914 21:57:06.101362 140584250015744 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
  WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
  W0914 21:57:06.101471 140584250015744 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
  INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  I0914 21:57:06.106059 140584250015744 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
  INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.01s
  I0914 21:57:06.106375 140584250015744 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.01s
  [  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
  ----------------------------------------------------------------------
  Ran 4 tests in 0.525s
  
  OK (skipped=4)
  ================================================================================
  Target //keras/distribute:multi_worker_test up-to-date:
    bazel-bin/keras/distribute/multi_worker_test
  INFO: Elapsed time: 6.094s, Critical Path: 5.49s
  INFO: 6 processes: 4 internal, 2 local.
  INFO: Build completed successfully, 6 total actions
  //keras/distribute:multi_worker_test                                     PASSED in 5.4s
    Stats over 2 runs: max = 5.4s, min = 5.4s, avg = 5.4s, dev = 0.0s
  
  Executed 1 out of 1 test: 1 test passes.
  INFO: Build completed successfully, 6 total actions
  ```
</details>

Reply:
Is the default training loop test case broken as well? If so, if we were to pick `reduction="first"` for multi-worker strategy, that could fix custom training loop test case, but could potentially break for the default one?
Spitballing here: Maybe the cleanest solution would be to always concatenate and then sum? I can see `_multi_worker_concat` has a little bit more logic to avoid repetitions and deal with different shapes.

Reply:
Yeah, we'll definitely need to understand why the tests are being skipped. The tests are using one gpu only but on multiple "simulated" workers. Can we try one thing: attempt to enclose the whole test method with a `try`-`except` to catch `unittest.SkipTest` (note that you would need to `import unittest`), and log the exception to see the content of it? It's unfortunate the test doesn't tell you why to skip which we should fix, but hopefully by catching it like this we get to see what's going on there. Thanks for your patience.

Reply:
Added the try catch block to setUp, setUpClass, the test method itself and created a new test that called it directly, like so:
```py
def test_something_i_guess(self):
        try:
            self.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU()
        except unittest.SkipTest as e:
            _print_skip(e)
            raise e
```

What I believe to be the important bit:
```shell
********************************************************************************
TODO(b/150264776): Resolve dependency issue in CI
********************************************************************************
Traceback (most recent call last):
  File "/home/ldavid/.cache/bazel/_bazel_ldavid/e04af3143c17f8e2b54fd8ec88c7bad1/execroot/org_keras/bazel-out/k8-opt/bin/keras/distribute/multi_worker_test.runfiles/org_keras/keras/distribute/multi_worker_test.py", line 219, in test_something_i_guess
    self.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU()
  File "/home/ldavid/.local/lib/python3.10/site-packages/absl/testing/parameterized.py", line 312, in bound_param_test
    return test_method(self, **testcase_params)
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/framework/test_combinations.py", line 360, in decorated
    execute_test_method()
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/framework/test_combinations.py", line 343, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/distribute/combinations.py", line 580, in decorator
    results = runner.run(_test_runner, args=(test_id, _env))
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/distribute/multi_process_runner.py", line 981, in run
    self._start()
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/distribute/multi_process_runner.py", line 944, in _start
    raise unittest.SkipTest(
unittest.case.SkipTest: TODO(b/150264776): Resolve dependency issue in CI
********************************************************************************
```

<details>
<summary>full logs</summary>

```shell
WARNING: The following configs were expanded more than once: [v2]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=207
INFO: Reading rc options for 'test' from /home/ldavid/workspace/repos/keras/.bazelrc:
  Inherited 'build' options: --apple_platform_type=macos --define open_source_build=true --define=use_fast_cpp_protos=false --define=tensorflow_enable_mlir_generated_gpu_kernels=0 --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --config=short_logs --config=v2
INFO: Reading rc options for 'test' from /home/ldavid/workspace/repos/keras/.bazelrc:
  'test' options: --define open_source_build=true --define=use_fast_cpp_protos=false --config=v2
INFO: Found applicable config definition build:short_logs in file /home/ldavid/workspace/repos/keras/.bazelrc: --output_filter=DONT_MATCH_ANYTHING
INFO: Found applicable config definition build:v2 in file /home/ldavid/workspace/repos/keras/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Found applicable config definition build:v2 in file /home/ldavid/workspace/repos/keras/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Analyzed target //keras/distribute:multi_worker_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
INFO: From Testing //keras/distribute:multi_worker_test (shard 1 of 2):
==================== Test output for //keras/distribute:multi_worker_test (shard 1 of 2):
2022-09-15 02:57:57.667868: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-15 02:57:58.493165: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.7/lib64:/usr/local/cuda/extras/CUPTI/lib64
2022-09-15 02:57:58.493236: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.7/lib64:/usr/local/cuda/extras/CUPTI/lib64
2022-09-15 02:57:58.493244: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
Running tests under Python 3.10.4: /home/ldavid/workspace/envs/k/local/bin/python3
[ RUN      ] KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False
2022-09-15 02:57:59.369153: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-15 02:57:59.835460: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:42] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2022-09-15 02:57:59.835504: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6767 MB memory:  -> device: 0, name: NVIDIA GeForce GTX 1070 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1
INFO:tensorflow:Using local port 41773
I0915 02:57:59.836780 139627678822400 test_util.py:3743] Using local port 41773
INFO:tensorflow:Using local port 36527
I0915 02:57:59.837091 139627678822400 test_util.py:3743] Using local port 36527
W0915 02:57:59.837304 139627678822400 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.837437 139627678822400 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.854940 139627678822400 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False): 0.49s
I0915 02:57:59.855267 139627678822400 test_util.py:2458] time(__main__.KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False): 0.49s
[  SKIPPED ] KPLMultiWorkerTest.testTrainAndServeWithKPL_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU_useadapt_False
[ RUN      ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
W0915 02:57:59.855844 139627678822400 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.855946 139627678822400 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.858248 139627678822400 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
I0915 02:57:59.858485 139627678822400 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
[ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
W0915 02:57:59.858984 139627678822400 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.859076 139627678822400 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.861384 139627678822400 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
I0915 02:57:59.861634 139627678822400 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
[ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
W0915 02:57:59.862103 139627678822400 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.862202 139627678822400 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.864597 139627678822400 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:CPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
I0915 02:57:59.864848 139627678822400 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU): 0.0s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1CPU
[ RUN      ] KerasMultiWorkerTestIndependentWorker.test_session
********************************************************************************
Not a test.
********************************************************************************
Traceback (most recent call last):
  File "/home/ldavid/.cache/bazel/_bazel_ldavid/e04af3143c17f8e2b54fd8ec88c7bad1/execroot/org_keras/bazel-out/k8-opt/bin/keras/distribute/multi_worker_test.runfiles/org_keras/keras/distribute/multi_worker_test.py", line 212, in setUp
    return super().setUp()
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/framework/test_util.py", line 2450, in setUp
    self.skipTest("Not a test.")
  File "/usr/lib/python3.10/unittest/case.py", line 671, in skipTest
    raise SkipTest(reason)
unittest.case.SkipTest: Not a test.
********************************************************************************
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_session
----------------------------------------------------------------------
Ran 5 tests in 0.498s

OK (skipped=5)
================================================================================
INFO: From Testing //keras/distribute:multi_worker_test (shard 2 of 2):
==================== Test output for //keras/distribute:multi_worker_test (shard 2 of 2):
2022-09-15 02:57:57.660809: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-15 02:57:58.523730: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.7/lib64:/usr/local/cuda/extras/CUPTI/lib64
2022-09-15 02:57:58.523799: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: :/usr/local/cuda-11.7/lib64:/usr/local/cuda/extras/CUPTI/lib64
2022-09-15 02:57:58.523805: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
Running tests under Python 3.10.4: /home/ldavid/workspace/envs/k/local/bin/python3
[ RUN      ] KPLMultiWorkerTest.test_session
[  SKIPPED ] KPLMultiWorkerTest.test_session
[ RUN      ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
2022-09-15 02:57:59.384726: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-15 02:57:59.836991: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:42] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2022-09-15 02:57:59.837030: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6767 MB memory:  -> device: 0, name: NVIDIA GeForce GTX 1070 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1
INFO:tensorflow:Using local port 38279
I0915 02:57:59.838176 140395605262336 test_util.py:3743] Using local port 38279
INFO:tensorflow:Using local port 38063
I0915 02:57:59.838501 140395605262336 test_util.py:3743] Using local port 38063
W0915 02:57:59.838683 140395605262336 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.838810 140395605262336 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.856771 140395605262336 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.47s
I0915 02:57:59.857095 140395605262336 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.47s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.testSimpleModelIndependentWorkerSync_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
[ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
W0915 02:57:59.857683 140395605262336 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.857789 140395605262336 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.861453 140395605262336 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.0s
I0915 02:57:59.861694 140395605262336 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.0s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_custom_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
[ RUN      ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
W0915 02:57:59.862202 140395605262336 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.862303 140395605262336 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.865835 140395605262336 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.0s
I0915 02:57:59.866064 140395605262336 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU): 0.0s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU
[ RUN      ] KerasMultiWorkerTestIndependentWorker.test_something_i_guess
W0915 02:57:59.866637 140395605262336 context.py:746] Configuring coordination service type may not be effective because the context is already initialized.
WARNING:tensorflow:Collective ops is not configured at program startup. Some performance features may not be enabled.
W0915 02:57:59.866742 140395605262336 collective_all_reduce_strategy.py:394] Collective ops is not configured at program startup. Some performance features may not be enabled.
INFO:tensorflow:Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
I0915 02:57:59.870365 140395605262336 collective_all_reduce_strategy.py:446] Single-worker MultiWorkerMirroredStrategy with local_devices = ('/device:GPU:0',), communication = CommunicationImplementation.AUTO
********************************************************************************
TODO(b/150264776): Resolve dependency issue in CI
********************************************************************************
Traceback (most recent call last):
  File "/home/ldavid/.cache/bazel/_bazel_ldavid/e04af3143c17f8e2b54fd8ec88c7bad1/execroot/org_keras/bazel-out/k8-opt/bin/keras/distribute/multi_worker_test.runfiles/org_keras/keras/distribute/multi_worker_test.py", line 219, in test_something_i_guess
    self.test_distribution_reduction_method_auto_default_train_step_test_mode_eager_strategy_MultiWorkerMirrored2x1GPU()
  File "/home/ldavid/.local/lib/python3.10/site-packages/absl/testing/parameterized.py", line 312, in bound_param_test
    return test_method(self, **testcase_params)
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/framework/test_combinations.py", line 360, in decorated
    execute_test_method()
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/framework/test_combinations.py", line 343, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/distribute/combinations.py", line 580, in decorator
    results = runner.run(_test_runner, args=(test_id, _env))
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/distribute/multi_process_runner.py", line 981, in run
    self._start()
  File "/home/ldavid/.local/lib/python3.10/site-packages/tensorflow/python/distribute/multi_process_runner.py", line 944, in _start
    raise unittest.SkipTest(
unittest.case.SkipTest: TODO(b/150264776): Resolve dependency issue in CI
********************************************************************************
INFO:tensorflow:time(__main__.KerasMultiWorkerTestIndependentWorker.test_something_i_guess): 0.0s
I0915 02:57:59.871345 140395605262336 test_util.py:2458] time(__main__.KerasMultiWorkerTestIndependentWorker.test_something_i_guess): 0.0s
[  SKIPPED ] KerasMultiWorkerTestIndependentWorker.test_something_i_guess
----------------------------------------------------------------------
Ran 5 tests in 0.488s

OK (skipped=5)
================================================================================
Target //keras/distribute:multi_worker_test up-to-date:
  bazel-bin/keras/distribute/multi_worker_test
INFO: Elapsed time: 2.631s, Critical Path: 2.54s
INFO: 3 processes: 1 internal, 2 local.
INFO: Build completed successfully, 3 total actions
//keras/distribute:multi_worker_test                                     PASSED in 2.5s
  Stats over 2 runs: max = 2.5s, min = 2.5s, avg = 2.5s, dev = 0.0s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 3 total actions
```
</details>

Reply:
Thanks! This is very useful information. It's basically [this](https://github.com/tensorflow/tensorflow/blob/349438ddeae7b03a6c5aaca6e731a998c6f75965/tensorflow/python/distribute/multi_process_runner.py#L945).

MultiProcessRunner requires `dill` to run. Would you mind trying installing `dill` package and see if it helps?

Reply:
Yep, its working now. (In the sense that the test case is broken).
Thank you for the help. I'll try to find a solution for this.

Reply:
Great, thanks. So just to confirm, you were able to reproduce with a log result that I pasted earlier?

Reply:
Yes.